builder-oapi.yaml

```
@@ -0,0 +1,74 @@
+openapi: "3.0.3"
+
+info:
+  title: "Builder API"
+  description: |
+    API specification for external builder clients, which enables validators to delegate block building duties.
+
+    All requests by default send and receive JSON, and as such should have either or both of the "Content-Type: application/json"
+    and "Accept: application/json" headers. In addition, some requests can return data in the SSZ format. To indicate that SSZ
+    data is required in response to a request the header "Accept: application/octet-stream" should be sent. Note that only a subset
+    of requests can respond with data in SSZ format; these are noted in each individual request.
+
+    API endpoints are individually versioned.  As such, there is no direct relationship between all v1 endpoints, all v2 endpoints,
+    _etc._ and no such relationship should be inferred.  All JSON responses return the requested data under a `data` key in the top
+    level of their response.  Additional metadata may or may not be present in other keys at the top level of the response, dependent
+    on the endpoint.  The rules that require an increase in version number are as follows:
+
+      - no field that is listed in an endpoint shall be removed without an increase in the version number
+      - no field that is listed in an endpoint shall be altered in terms of format (_e.g._ from a string to an array) without an
+        increase in the version number
+
+    Note that it is possible for a field to be added to an endpoint's data or metadata without an increase in the version number.
+
+  version: "dev"
+  contact:
+    name: Ethereum Github
+    url: https://github.com/ethereum/builder-apis/issues
+  license:
+    name: "Apache 2.0"
+    url: "https://www.apache.org/licenses/LICENSE-2.0.html"
+
+servers:
+  - url: "{server_url}"
+    variables:
+      server_url:
+        description: "Builder node API url"
+        default: "???"
+
+tags:
+  - name: Builder
+    description: Set of endpoints to interact with an external block builder.
+paths:
+
+  /eth/v1/builder/validators:
+    $ref: "./apis/builder/validators.yaml"
+  /eth/v1/builder/{slot}/{parent_hash}/header:
```

missing pubkey?

```suggestion
  /eth/v1/builder/{slot}/{parent_hash}/{pubkey}/header:
```


could also consider query params or moving this to a `POST` with the request data

---

`pubkey` is currently listed as a query parameter. I don't have a good feel for the proper formatting for this method. Originally, it felt like it was more aligned with `GET`, but after more contemplation `PUT` feels a bit more appropriate because the request _might_ cause the builder to generate a response on a different `parent_hash`. I think `PUT` is better than `POST` here because this method is mostly idempotent, because the act of calling the method does not affect the result. However, it is possible for the result to change because the builder may find a better block between requests. Since that is completely separate from the request, I think it's okay though.

---

I think it would be best to have header first, like this:
```/eth/v1/builder/header```

Maybe put all arguments as query args, rather than some in the path and others as query args? 🤔 

I guess any method works, but imo GET seems fine because it will not modify anything on the relay. The relay will just return whatever block it has for these settings or none. This call will probably not be forwarded to the builder anyways (when we were thinking through relay+builder interaction, it seems natural for builders to stream new blocks to the relay, mostly for timing reasons).

---

I will stick with `GET` for now then, but I do think path vars are better because query parameters typically filter results whereas we need all three values to specify the resource we're requesting.